rust/src/smb/smb1_records.rs

```diff
@@ -0,0 +1,647 @@
+/* Copyright (C) 2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use log::*;
+use nom::{rest, le_u8, le_u16, le_u32, le_u64, IResult};
+
+#[derive(Debug,PartialEq)]
+pub struct Smb1WriteRequestRecord<'a> {
+    pub offset: u64,
+    pub len: u32,
+    pub fid: &'a[u8],
+    pub data: &'a[u8],
+}
+
+named!(pub parse_smb1_write_request_record<Smb1WriteRequestRecord>,
+    do_parse!(
+            wct: le_u8
+        >>  fid: take!(2)
+        >>  count: le_u16
+        >>  offset: le_u32
+        >>  remaining: le_u16
+        >>  bcc: le_u16
+        >>  buffer_format: le_u8
+        >>  data_len: le_u16
+        >>  file_data: take!(data_len)
+        >> (Smb1WriteRequestRecord {
+                offset: offset as u64,
+                len: data_len as u32,
+                fid:fid,
+                data:file_data,
+            }))
+);
+
+named!(pub parse_smb1_write_andx_request_record<Smb1WriteRequestRecord>,
+    do_parse!(
+            wct: le_u8
+        >>  andx_command: le_u8
+        >>  take!(1)    // reserved
+        >>  andx_offset: le_u16
+        >>  fid: take!(2)
+        >>  offset: le_u32
+        >>  take!(4)    // reserved
+        >>  write_mode: le_u16
+        >>  remaining: le_u16
+        >>  data_len_high: le_u16
+        >>  data_len_low: le_u16
+        >>  data_offset: le_u16
+        >>  high_offset: cond!(wct==14,le_u32)
+        >>  bcc: le_u16
+        //>>  padding: cond!(data_offset > 32, take!(data_offset - 32))
+        >>  padding: cond!(bcc > data_len_low, take!(bcc - data_len_low)) // TODO figure out how this works with data_len_high
+        >>  file_data: rest
+        >> (Smb1WriteRequestRecord {
+                offset: if high_offset != None { ((high_offset.unwrap() as u64) << 32)|(offset as u64) } else { 0 },
```

Comment:
maybe `unwrap_or` would simplify the above line

Reply:
I'm failing to get it to work. Any suggestions to how?